### ISSUE TITLE
/labelValue?filter multi-select filtering for Grafana

### ISSUE THREAD

**Issue by willr3** [Hyperfoil/Horreum#1695]
filters were created for 2 use cases:
1. Does the provided filter object exist in the label value map? This allows user to pass in a "subset" of the label value map and filter for only those maps that match. This is akin to passing in a fingerprint object and only getting back label value maps that contain the fingerprint.
2. Does the provided jsonpath filter match the label value map? e.g. `$."Number of Samples" ? (@ > 50)`. This is a more powerful option but more complicated and is directly tied to the implementation of our jsonpath processing (postgres) so I do not think we should advertise this option to users atm.

Option 1 works with grafana filtering when looking for specific fingerprints. It does NOT work with grafana variable based fintering in 2 situations:
1. Selecting all values.
if label "cpu" has possible values 1,2,4,8 then grafana's default option to select all values using variables would be in included `"cpu": [1,2,4,8]` (assuming json encoding). Horreum would see this as a label with an array value and not an array of possible values
Grafana offers custom encoding for "All" values (e.g. `*`) but at the moment Horreum would see that a string value for the "cpu" label.
Horreum's current method to select All values would be to exclude the label name from the fingerprint. Unfortunately Grafana does not appear to offer that option when configuring filtering.
2. Selecting multiple values.
Similar to the default all values, Grafana will encode selecting some of the values as an array of options. e.g. `"cpu": [4,8]`. Horreum would see this as a label with an array value and not an array of possible values.

Options:

I believe Horreum must support Grafana multi-select if it will be considered a viable solution for user charting.

* We can support the "all values" problem relatively easily by checking for a custom encoded value (e.g. `*`) and removing that from the filter object. This will prevent all label's from using that value as a valid label value. I don't like the idea of doing this but it is admittedly easier and faster to implement. 
The problem is that this is a "Forever" decision or a breaking api change whenever / if ever we revert it. I personally think we should not introduce this option because it is only a partial solution (doesn't address multiple value selection) and it places value restrictions on all labels.

* We could similarly use graphite encoding ( `"cpu": {1,2,4,8}`) but inside a string ( `"cpu": "{1,2,4,8}"` ) and use the same custom filter from above to determine this is an array of options and not the desired content of a string. This along with some indicator paramater to "enable grafana parameter parsing" would be able to support both all and multi-select. One downside is that it removes the possible label value (probably a minor detail) and it mixes model representation. I am not a fan of including json and grafite because it is tightly coupled with grafana. The next UI charting tool (jupyter, tablua, etc) would have a different query syntax and we would end up with a growing list of integration support.

The next option is to introduce a domain specific filtering syntax that is compatible with Grafana. This would need to support how Grafana encodes single `"cpu": 4`, multiple `"cpu": [1,8]`, and all variable options `"cpu" : [1,2,4,8]` or `"cpu":"*"`. 
The challenge is to differentiate arrays that represent value options versus label values that are arrays. 

I looked at the domain languages used by mongo or elastic but they both make use of Grafana plugins to convert the variable input into their query syntax.

A naive solution to this could be to perform both queries and return the longest list of results. The theory being the incorrect interpretation would return 0 results (checking for a equality filter object is a multi-select vs checking for containment when the label value is an array)
This is not a good choice because it requires 2x the number of arrays in the filter to find the correct solution. That will hopefully usually be small but is still quite wasteful.

**Comment by johnaohara** [comment]
> This would run the label value lookup N+1 times where N is the number of top level arrays in the filter object. Horreum could make further performance improvements with some caching but the N+1 checks would remain.

Why would we have the N+1 problem?  

If we had more than one top level filter, say `cpu:[1,8]` and `region:["apac", "emea"]`, wouldn't the pseudo-query become;

```
select ....
where
   ... AND `cpu` in (1, 8) and `region` in ("apac", "emea")
```

**Comment by johnaohara** [comment]
> Why would we have the N+1 problem?

Sry, read comment before coffee, N+1 comes from the fact we need to scan first to see if sub object exists.

**Comment by johnaohara** [comment]
I like the idea of **not** providing grafite dialect support directly into the api.

To determine if arrays are arrays of value options vs label values, could we not construct a single query that checks all top level arrays and returns a row containing the result for all arrays in the query? Maybe returning a count, 

e.g. `cpu: [1,8]` and `regions: ["emea", "apac"]`, a single query that returns;

| cpu  | regions |
| ------------- | ------------- |
| 0  | 20  |

The we assume that `cpu` is an array of options, whereas `regions` is a label value?

**Comment by willr3** [comment]
Yes, we could create a sql that performs all N checks in a single query. It would be a massive query but that isn't new for Horreum.

I have an implementation working with the N+1 approach so we know the multi-select filtering is feasible. Hibernate was originally keen on turning a parameter into multiple parameters in the sql which didn't work with the postgres type checking. 

Do we want to have array checking for multi-select on by default or require another query parameter to enable it (e.g. `?multiSelect=true`) ?  Having it on by default avoids changing the openApi spec but introduces overhead to all `/labelValues` requests that include an array in the `filter` parameter. 

I will look at the complexity of a single sql query but I think the sql will eventually change when Horreum starts to track object "shape" for run data and label values.

**Comment by johnaohara** [comment]
I wouldn't worry about the openAPI spec, if it changes, we will just bump to 0.14.  We need an api that makes the most sense rationally and does not complicate the query parsing logic
